File: db4oj/core/src/com/db4o/internal/btree/BTreeUpdate.java
/* Copyright (C) 2006 db4objects Inc. http://www.db4o.com */

package com.db4o.internal.btree;

/**
 * @exclude
 */
import com.db4o.foundation.No4;
import com.db4o.internal.*;

public abstract class BTreeUpdate extends BTreePatch {

	protected BTreeUpdate _next;

	public BTreeUpdate(Transaction transaction, Object obj) {
		super(transaction, obj);
	}

	protected boolean hasNext() {
		return _next != null;
	}

	public BTreePatch forTransaction(Transaction trans) {
	    if(_transaction == trans){
	        return this;
	    }
	    if(_next == null){
	        return null;
	    }
	    return _next.forTransaction(trans);
	}

	public BTreeUpdate removeFor(Transaction trans) {
		if (_transaction == trans) {
			return _next;
		}
		if (_next == null) {
			return this;
		}
		return _next.removeFor(trans);
	}

	public void append(BTreeUpdate patch) {
	    if(_transaction == patch._transaction){
	    	// don't allow two patches for the same transaction
	        throw new IllegalArgumentException();
	    }
	    if(!hasNext()){
	        _next = patch;
	    }else{
	        _next.append(patch);
	    }
	}
	
    protected void applyKeyChange(Object obj) {
        _object = obj;
        if (hasNext()) {
            _next.applyKeyChange(obj);      
        }
    }

	protected abstract void committed(BTree btree);

	public Object commit(Transaction trans, BTree btree) {
		final BTreeUpdate patch = (BTreeUpdate) forTransaction(trans);
		if (patch instanceof BTreeCancelledRemoval) {
			Object obj = patch.getCommittedObject();
			applyKeyChange(obj);
		} 
	    return internalCommit(trans, btree);
	}

	protected Object internalCommit(Transaction trans, BTree btree) {
		if(_transaction == trans){	        
	        committed(btree);
	        if (hasNext()){
	            return _next;
	        }
	        return getCommittedObject();
	    }
	    if(hasNext()){
	        setNextIfPatch(_next.internalCommit(trans, btree));
	    }
	    return this;
	}

	private void setNextIfPatch(Object newNext) {
		if(newNext instanceof BTreeUpdate){
			_next = (BTreeUpdate)newNext;
		} else {
		    _next = null;
		}
	}

	protected abstract Object getCommittedObject();

	public Object rollback(Transaction trans, BTree btree) {
	    if(_transaction == trans){
	        if(hasNext()){
	            return _next;
	        }
	        return getObject();
	    }
	    if(hasNext()){
	        setNextIfPatch(_next.rollback(trans, btree));
	    }
	    return this;
	}
	
	public Object key(Transaction trans) {
		BTreePatch patch = forTransaction(trans);
		if (patch == null) {
			return getObject();
		}
		if (patch.isRemove()) {
			return No4.INSTANCE;
		}
		return patch.getObject();
	}

}

File: db4on/core/src/core/com/db4o/internal/btree/BTreeUpdate.cs
namespace com.db4o.@internal.btree
{
	public abstract class BTreeUpdate : com.db4o.@internal.btree.BTreePatch
	{
		protected com.db4o.@internal.btree.BTreeUpdate _next;

		public BTreeUpdate(com.db4o.@internal.Transaction transaction, object obj) : base
			(transaction, obj)
		{
		}

		protected virtual bool HasNext()
		{
			return _next != null;
		}

		public override com.db4o.@internal.btree.BTreePatch ForTransaction(com.db4o.@internal.Transaction
			 trans)
		{
			if (_transaction == trans)
			{
				return this;
			}
			if (_next == null)
			{
				return null;
			}
			return _next.ForTransaction(trans);
		}

		public virtual com.db4o.@internal.btree.BTreeUpdate RemoveFor(com.db4o.@internal.Transaction
			 trans)
		{
			if (_transaction == trans)
			{
				return _next;
			}
			if (_next == null)
			{
				return this;
			}
			return _next.RemoveFor(trans);
		}

		public virtual void Append(com.db4o.@internal.btree.BTreeUpdate patch)
		{
			if (_transaction == patch._transaction)
			{
				throw new System.ArgumentException();
			}
			if (!HasNext())
			{
				_next = patch;
			}
			else
			{
				_next.Append(patch);
			}
		}

		protected virtual void ApplyKeyChange(object obj)
		{
			_object = obj;
			if (HasNext())
			{
				_next.ApplyKeyChange(obj);
			}
		}

		protected abstract void Committed(com.db4o.@internal.btree.BTree btree);

		public override object Commit(com.db4o.@internal.Transaction trans, com.db4o.@internal.btree.BTree
			 btree)
		{
			com.db4o.@internal.btree.BTreeUpdate patch = (com.db4o.@internal.btree.BTreeUpdate
				)ForTransaction(trans);
			if (patch is com.db4o.@internal.btree.BTreeCancelledRemoval)
			{
				object obj = patch.GetCommittedObject();
				ApplyKeyChange(obj);
			}
			return InternalCommit(trans, btree);
		}

		protected virtual object InternalCommit(com.db4o.@internal.Transaction trans, com.db4o.@internal.btree.BTree
			 btree)
		{
			if (_transaction == trans)
			{
				Committed(btree);
				if (HasNext())
				{
					return _next;
				}
				return GetCommittedObject();
			}
			if (HasNext())
			{
				SetNextIfPatch(_next.InternalCommit(trans, btree));
			}
			return this;
		}

		private void SetNextIfPatch(object newNext)
		{
			if (newNext is com.db4o.@internal.btree.BTreeUpdate)
			{
				_next = (com.db4o.@internal.btree.BTreeUpdate)newNext;
			}
			else
			{
				_next = null;
			}
		}

		protected abstract object GetCommittedObject();

		public override object Rollback(com.db4o.@internal.Transaction trans, com.db4o.@internal.btree.BTree
			 btree)
		{
			if (_transaction == trans)
			{
				if (HasNext())
				{
					return _next;
				}
				return GetObject();
			}
			if (HasNext())
			{
				SetNextIfPatch(_next.Rollback(trans, btree));
			}
			return this;
		}

		public override object Key(com.db4o.@internal.Transaction trans)
		{
			com.db4o.@internal.btree.BTreePatch patch = ForTransaction(trans);
			if (patch == null)
			{
				return GetObject();
			}
			if (patch.IsRemove())
			{
				return com.db4o.foundation.No4.INSTANCE;
			}
			return patch.GetObject();
		}
	}
}
